### PR TITLE
Add LuaCheck workflow

### PR DESCRIPTION
**Pull request by OgelGames**
I was able to fix most of the luacheck warnings, but there are a bunch of global variables that seem to be for configuration, that I have no clue what to do with:

```
dryplants\init.lua:22:4: accessing undefined variable 'REEDMACE_GENERATES'
dryplants\init.lua:25:4: accessing undefined variable 'SMALL_JUNCUS_GENERATES'
dryplants\init.lua:28:4: accessing undefined variable 'EXTRA_TALL_GRASS_GENERATES'
dryplants\init.lua:147:13: accessing undefined variable 'HAY_DRYING_TIME'
dryplants\init.lua:192:13: accessing undefined variable 'GRASS_REGROWING_TIME'
dryplants\init.lua:193:15: accessing undefined variable 'GRASS_REGROWING_CHANCE'

dryplants\juncus.lua:104:17: accessing undefined variable 'JUNCUS_NEAR_WATER_PER_MAPBLOCK'
dryplants\juncus.lua:105:20: accessing undefined variable 'JUNCUS_NEAR_WATER_RARITY'
dryplants\juncus.lua:125:17: accessing undefined variable 'JUNCUS_AT_BEACH_PER_MAPBLOCK'
dryplants\juncus.lua:126:20: accessing undefined variable 'JUNCUS_AT_BEACH_RARITY'

dryplants\moregrass.lua:18:15: accessing undefined variable 'TALL_GRASS_PER_MAPBLOCK'
dryplants\moregrass.lua:19:18: accessing undefined variable 'TALL_GRASS_RARITY'

dryplants\reed.lua:78:4: accessing undefined variable 'AUTO_ROOF_CORNER'
dryplants\reed.lua:236:4: accessing undefined variable 'REED_WILL_DRY'
dryplants\reed.lua:253:15: accessing undefined variable 'REED_DRYING_TIME'

dryplants\reedmace.lua:278:13: accessing undefined variable 'REEDMACE_GROWING_TIME'
dryplants\reedmace.lua:279:15: accessing undefined variable 'REEDMACE_GROWING_CHANCE'
dryplants\reedmace.lua:359:17: accessing undefined variable 'REEDMACE_NEAR_WATER_PER_MAPBLOCK'
dryplants\reedmace.lua:360:20: accessing undefined variable 'REEDMACE_NEAR_WATER_RARITY'
dryplants\reedmace.lua:382:17: accessing undefined variable 'REEDMACE_IN_WATER_PER_MAPBLOCK'
dryplants\reedmace.lua:383:17: accessing undefined variable 'REEDMACE_IN_WATER_RARITY'
dryplants\reedmace.lua:400:17: accessing undefined variable 'REEDMACE_FOR_OASES_PER_MAPBLOCK'
dryplants\reedmace.lua:401:20: accessing undefined variable 'REEDMACE_FOR_OASES_RARITY'

trunks\generating.lua:33:5: accessing undefined variable 'Big_Twigs'
trunks\generating.lua:154:4: accessing undefined variable 'Twigs_on_ground'
trunks\generating.lua:157:17: accessing undefined variable 'Twigs_on_ground_Max_Count'
trunks\generating.lua:158:14: accessing undefined variable 'Twigs_on_ground_Rarity'
trunks\generating.lua:171:4: accessing undefined variable 'Twigs_on_water'
trunks\generating.lua:174:17: accessing undefined variable 'Twigs_on_water_Max_Count'
trunks\generating.lua:175:14: accessing undefined variable 'Twigs_on_water_Rarity'
trunks\generating.lua:214:4: accessing undefined variable 'Horizontal_Trunks'
trunks\generating.lua:257:32: accessing undefined variable 'Horizontal_Trunks'
trunks\generating.lua:289:32: accessing undefined variable 'Horizontal_Trunks'
trunks\generating.lua:327:17: accessing undefined variable 'Trunks_Max_Count'
trunks\generating.lua:328:14: accessing undefined variable 'Trunks_Rarity'
trunks\generating.lua:345:4: accessing undefined variable 'Moss_on_ground'
trunks\generating.lua:361:17: accessing undefined variable 'Moss_on_ground_Max_Count'
trunks\generating.lua:362:14: accessing undefined variable 'Moss_on_ground_Rarity'
trunks\generating.lua:383:4: accessing undefined variable 'Moss_on_trunk'
trunks\generating.lua:461:17: accessing undefined variable 'Moss_on_trunk_Max_Count'
trunks\generating.lua:462:14: accessing undefined variable 'Moss_on_trunk_Rarity'
trunks\generating.lua:475:4: accessing undefined variable 'Roots'

trunks\nodes.lua:239:4: accessing undefined variable 'Auto_Roof_Corner'
trunks\nodes.lua:329:4: accessing undefined variable 'Roots'
```

Also line lengths are ignored for now.

EDIT: they are for configuration, possibly a relic from before `minetest.settings` existed?
https://github.com/mt-mods/plantlife_modpack/blob/master/dryplants/settings.txt
https://github.com/mt-mods/plantlife_modpack/blob/master/trunks/trunks_settings.txt

### CODE REVIEW

**Comment by JordanL2** [comment]
For `dryplants` and `trunks`, those settings declared in `settings.txt` and `trunk_settings.txt` need to move into `settingtypes.txt` files. I don't see any other mods in the modpack with those kind of settings files.

**Comment by Panquesito7** [comment]
Is this still going to be worked on?

**Comment by OgelGames** [comment]
> Is this still going to be worked on?

Not sure, I haven't really got time to work on it at the moment. Feel free to do whatever you want with it :)

**Comment by Panquesito7** [comment]
Working on this. I believe all we need is to ignore the old settings variables, right?

**Comment by OgelGames** [comment]
Preferably the old settings should be moved to `settingtypes.txt`, but you could just remove the settings altogether.

**Comment by Niklp09** [comment]
I've fixed all Luacheck warnigs already in other PR's. All that remains are the old settings, luacheck workflow and luacheck globals config

**Comment by Panquesito7** [comment]
> Preferably the old settings should be moved to `settingtypes.txt`, but you could just remove the settings altogether.

Yeah, I think that would fit for another PR.

**Comment by S-S-X** [comment]
My personal opinion about ignoring tests with comment that tests should not be ignored:

Shouldn't ignore stuff that should be fixed, luacheck (like any test that is working correctly and setting some standards) IMO is supposed to fail as long as things have not been fixed.

New tests that are still failing can be added and merged, then it is clear that code is not up to the standards we want but it does not prevent working on it.

Tests are supposed to tell you if there's still some important stuff to do and fix.
Adding tests just to have tests and then doing whatever it takes to keep green check marks even if it means ignoring all errors is very wrong thing to do and completely undermines actual purpose of tests.

**Comment by Panquesito7** [comment]
Should I fix the settings errors, then? I think that should be fixed in a separate PR, once fixed, we can merge this one.

**Comment by SwissalpS** [comment]
> Should I fix the settings errors, then? I think that should be fixed in a separate PR, once fixed, we can merge this one.

The title of this PR is "Add luacheck workflow and fix warnings" so I'm confused why you don't want to fix the warnings in this PR.

**Comment by S-S-X** [comment]
> Should I fix the settings errors, then? I think that should be fixed in a separate PR, once fixed, we can merge this one.

Personally I see 3 options that are all valid:
* Remove ignores, fix problems and merge with check mark ✔️
* Update PR title, remove ignores, do not fix problems and merge with x mark ❌
* Remove ignores and leave PR open waiting for someone to fix problems.

I mean merging new tests that are failing shouldn't be a problem if tests are working correctly: they are failing if there's still problems.

Then later in different PR someone can **fix** problems instead of "fixing" tests.

In other words: If tests are failing correctly, do not attempt to fix them. If PR fixes some issues it is still better than not fixing any issues. Green check mark shouldn't be needed for merge if there's old problems that will not be fixed in same PR, it can be made requirement later when another PR switches ❌ to ✔️.

**Comment by Panquesito7** [comment]
I think I'll leave it to someone else (option 2) for now. I don't have much time currently.

**Comment by wsor4035** [comment]
this pr is essentially the same as when it closed...... ill look at actually fixing the settings this weekend.

**Comment by Panquesito7** [comment]
> this pr is essentially the same as when it closed...... ill look at actually fixing the settings this weekend.

Technically yes 😅 if you need any help, let me know.

**Comment by wsor4035** [comment]
https://github.com/mt-mods/plantlife_modpack/pull/53
